mcp/tool_example_test.go

```diff
@@ -83,6 +83,122 @@ func ExampleAddTool_customTypeSchema() {
+	Cloudy       WeatherType = "clouds"
+	Rainy        WeatherType = "rain"
+	Snowy        WeatherType = "snow"
+)
+
+type Probability float64
+
+type WeatherOutput struct {
+	Summary       string      `json:"summary" jsonschema:"a summary of the weather forecast"`
+	Confidence    Probability `json:"confidence" jsonschema:"confidence, between 0 and 1"`
+	AsOf          time.Time   `json:"asOf" jsonschema:"the time the weather was computed"`
+	DailyForecast []Forecast  `json:"dailyForecast" jsonschema:"the daily forecast"`
+	Source        string      `json:"source,omitempty" jsonschema:"the organization providing the weather forecast"`
+}
+
+func WeatherTool(ctx context.Context, req *mcp.CallToolRequest, in WeatherInput) (*mcp.CallToolResult, WeatherOutput, error) {
+	perfectWeather := WeatherOutput{
+		Summary:    "perfect",
+		Confidence: 1.0,
+		AsOf:       time.Now(),
+	}
+	for range in.Days {
+		perfectWeather.DailyForecast = append(perfectWeather.DailyForecast, Forecast{
+			Forecast: "another perfect day",
+			Type:     Sunny,
+			Rain:     0.0,
+			High:     72.0,
+			Low:      72.0,
+		})
+	}
+	return nil, perfectWeather, nil
+}
+
+func ExampleAddTool_complexSchema() {
+	// This example demonstrates a tool with a more 'realistic' input and output
+	// schema. We use a combination of techniques to tune our input and output
+	// schemas.
+
+	// Distinguished Go types allow custom schemas to be reused during inference.
+	customSchemas := map[any]*jsonschema.Schema{
+		Probability(0):  {Type: "number", Minimum: jsonschema.Ptr(0.0), Maximum: jsonschema.Ptr(1.0)},
+		WeatherType(""): {Type: "string", Enum: []any{Sunny, PartlyCloudy, Cloudy, Rainy, Snowy}},
+	}
+	opts := &jsonschema.ForOptions{TypeSchemas: customSchemas}
+	in, err := jsonschema.For[WeatherInput](opts)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Furthermore, we can tweak the inferred schema, in this case limiting
+	// forecasts to 0-10 days.
+	daysSchema := in.Properties["days"]
+	daysSchema.Minimum = jsonschema.Ptr(0.0)
+	daysSchema.Maximum = jsonschema.Ptr(10.0)
+
+	// Output schema inference can reuse our custom schemas from input inference.
+	out, err := jsonschema.For[WeatherOutput](opts)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Now add our tool to a server. Since we've customized the schemas, we need
+	// to override the default schema inference.
+	server := mcp.NewServer(&mcp.Implementation{Name: "server", Version: "v0.0.1"}, nil)
+	mcp.AddTool(server, &mcp.Tool{
+		Name:         "weather",
+		InputSchema:  in,
+		OutputSchema: out,
+	}, WeatherTool)
+
+	ctx := context.Background()
+	session, err := connect(ctx, server) // create an in-memory connection
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer session.Close()
+
+	// Check that the client observes the correct schemas.
+	for t, err := range session.Tools(ctx, nil) {
+		if err != nil {
+			log.Fatal(err)
+		}
+		fmt.Println("max days:", *t.InputSchema.Properties["days"].Maximum)
```

Comment:
The best way to print the whole schema is to marshal it and print the string.
Just an idea; you may prefer this.

Reply:
Yeah, that's what the example above this one does, but it's too much output for here.